### ISSUE TITLE
`orderBy` filtered count of related object

### ISSUE THREAD

**Issue by AaronBaldwin** [prisma/prisma#20838]
## Problem

I would like to sort a query by the count of a related object.  However, I want to only count related objects that are in a specific status.

`prismaClient.author.findMany({
  orderBy: { posts: { _count: 'desc'} }
})`
^ This will sort by post count, but what if I am soft-deleting posts such that they still exist in the database as a relation, but the application does not recognize them.

## Suggested solution
Allow an optional where clause.
`prismaClient.author.findMany({
  orderBy: { posts: { where: { deleted: false }, _count: 'desc'} }
})`

## Alternatives

As far as I know the only alternative is to use a raw query.. which is not really possible for us to swap out our complicated logic.

**Comment by jkomyno** [comment]
Hi @AaronBaldwin, could you please help us understanding a bit more about what you're asking?

From what I understood, given a schema like

```prisma
model Author {
  id    String @id @default(cuid())
  name  String
  posts Post[]
}

model Post {
  id        String  @id @default(cuid())
  title     String
  deleted   Boolean @default(false)
  author    Author? @relation(fields: [authorId], references: [id])
  authorId  String?
}
```

and a seed script like

```typescript
import { PrismaClient } from '@prisma/client'

async function main() {
  const prisma = new PrismaClient()

  await prisma.author.create({
    data: {
      name: 'John',
      posts: {
        createMany: {
          data: [
            {
              title: 'Post 1',
              deleted: false,
            },
            {
              title: 'Post 2',
              deleted: true,
            },
            {
              title: 'Post 3',
              deleted: false,
            }
          ]
        }
      }
    },
  })

  await prisma.author.create({
    data: {
      name: 'Tim',
      posts: {
        createMany: {
          data: [
            {
              title: 'Post 4',
              deleted: false,
            },
            {
              title: 'Post 5',
              deleted: true,
            },
            {
              title: 'Post 6',
              deleted: true,
            }
          ]
        }
      }
    },
  })
}

main()
```

you'd like to formulate a Prisma query that returns the same result set as

```sql
WITH preliminary_count AS (
	SELECT
		a.id,
		COUNT(p.id) AS post_count
	FROM "Author" a
	JOIN "Post" p
		ON p."authorId" = a."id"
	WHERE p.deleted = FALSE
	GROUP BY a.id
)
SELECT
        a.id AS author_id,
	a.name,
        p.id AS post_id,
	p.title,
	p.deleted
FROM "Author" a
JOIN "preliminary_count" pc
	ON pc."id" = a."id"
JOIN "Post" p
	ON p."authorId" = a."id"
WHERE p.deleted = FALSE
ORDER BY pc.post_count DESC
```

and if that's not currently possible, you'd propose a new syntax with an optional `where` clause within the `orderBy` block:

```typescript
prisma.author.findMany({ orderBy: { posts: { where: { deleted: false }, _count: 'desc'} } })
```

Is that right?

**Comment by AaronBaldwin** [comment]
@jkomyno yes!  I used the author/post count as an example since Prisma uses that in the docs.  I took your SQL and tweaked is slightly to test on my database for my use case and that is pretty much exactly what I want.

I want to fetch all of the "Author" information in this case and sort it by the pc.post_count descending exactly like you have done.  In my case Im looking at `status="ACTIVE" instead of `delete=FALSE` but it is the same thing.

The only thing I left out is that we are also using `take` to limit the returned results to 25 so in SQL there is also a `LIMIT 25`.  If we were not using take/LIMIT then we may have just sorted it programmatically after the fetch, but we are counting on the database to paginate and sort for us.

I would be open to the whatever Prisma syntax the Prisma team thinks is most consistent with the existing product, but I think some sort of `where` clause makes sense to filter the count of related objects.  What you posted is exactly what I suggested and it makes sense to me to express it that way:

`prisma.author.findMany({ orderBy: { posts: { where: { deleted: false }, _count: 'desc'} } })`

**Comment by gpadillakambda** [comment]
@AaronBaldwin  Did you find some workaround to handle it, excluding the raw query?

**Comment by joao-moonward** [comment]
I believe that this feature would be incredibly beneficial. It is something that I truly need

**Comment by leBolideur** [comment]
Any updates here?

**Comment by riley-worthington** [comment]
This would be a very useful feature! I have this use case as well.

**Comment by wvpelt** [comment]
Also in need of this feature.
Is this on the roadmap?
Using a raw query would take a lot of effort to maintain in my use case.

**Comment by davmik2601** [comment]
Any updates here @jkomyno ?   This would be a very useful feature !

**Comment by JClackett** [comment]
Would love this too

**Comment by kanedaki** [comment]
Would love this too

**Comment by zdychacek** [comment]
Any plans on this feature?

**Comment by ludralph** [comment]
Closing this in favour of [14598](https://github.com/prisma/prisma/issues/14598)
